lib/message_queue_processor.rb

```diff
@@ -0,0 +1,92 @@
+class MessageQueueProcessor
+  def process(message)
+    summary = process_message(message.payload)
+    message.ack
+    Rails.logger.info summary.to_json
+  end
+
+  # For the details of unpublishing types, the publishing-api docs are
+  # the canonical source:
+  # https://github.com/alphagov/publishing-api/blob/main/docs/model.md#unpublishing
+  def process_message(payload)
+    affected = SavedPage.where(content_id: payload["content_id"])
+    affected_count = affected.count
+
+    effect =
+      case payload["document_type"]
+      when "gone"
+        redirect_saved_pages(affected, find_alternative_path(payload))
+      when "redirect"
+        redirect_saved_pages(affected, find_alternative_path(payload))
+      when "vanish"
+        destroy_saved_pages(affected)
+      else
+        update_saved_pages(affected, payload)
+      end
+
+    {
+      type: payload["document_type"],
+      base_path: payload["base_path"],
+      content_id: payload["content_id"],
+      affected_pages: affected_count,
+      effect: effect,
+    }
+  end
+
+  def find_alternative_path(payload)
+    alternative_path = payload.dig("details", "alternative_path")
+    return alternative_path if alternative_path
+
+    exact_redirect = payload["redirects"]&.find { |r| r["type"] == "exact" && payload["base_path"] == r["path"] }
+    return exact_redirect["destination"] if exact_redirect
+
+    prefix_redirect = payload["redirects"]&.select { |r| r["type"] == "prefix" && payload["base_path"].start_with?(r["path"]) }&.max_by { |r| r["path"].length }
```

Comment:
Interested so out of a number of possible redirects here, we're filtering then ultimately selecting the longest by path?
Why the longest?

Reply:
That's the behaviour router uses.  Say you visit `/government/organisations/ministry-of-magic/spellcasting-policy`, it'll look for these routes in this order:

1. An exact route for `/government/organisations/ministry-of-magic/spellcasting-policy`
2. A prefix route for `/government/organisations/ministry-of-magic/spellcasting-policy`
3. A prefix route for `/government/organisations/ministry-of-magic`
4. A prefix route for `/government/organisations`

Reply:
makes sense, cheers for the explaination!